[7주차] Team 프로메사 권동욱 & 김서연 미션 제출합니다.

2025314242 opened the pull request:
# 마치며

## 👨‍💻 권동욱
- 다음 과제에서 로그인/회원가입 포함하여 과제 완성도를 올릴 예정입니다.
- 다음 과제에서 반응형 디자인을 적용할 예정입니다.

## 👩‍💻 김서연
- Zustand로 투표 유형(파트장/데모데이), 선택된 후보 등의 상태를 각 페이지에서 공유할 수 있도록 했습니다. 현재는 단순 라우팅 기준으로 페이지 흐름이 제어되지만, 추후 로그인 기능 도입 시 사용자 상태에 따라 접근을 제한하는 방식으로 개선할 계획입니다.
- 결과 페이지에서는 득표 수 데이터를 받아올 수 있도록 했고, 임시로 득표 수를 적용하여 정렬과 표시 형태를 구현해 놓았습니다. 
- 투표 제출 시 간단한 애니메이션과, 하단 Indicator를 누르면 Logout 모달이 등장하는 구조도 함께 구현했습니다.

# 각종 모음

## 배포 링크
https://next-vote-21th-promesa.vercel.app

## 기본 설정
https://pickle-alarm-35a.notion.site/Frontend-Default-Setting-1f29c22f061f80cc960be21f2284505a?pvs=74

## 폴더 구조 및 디자인 초안
https://pickle-alarm-35a.notion.site/7-8-1f39c22f061f801db81ef32ba493e9f0

# Key Questions
## Zod 스키마가 무엇인지, 어떻게 활용할 수 있는지 알아봅시다.
- 입력 검증 로직을 중앙 집중화하여 중복 제거
- 스키마 자체가 타입 추론 가능하기 때문에, ```z.infer<typeof schema>```로 타입 선언도 자동화
- 런타임 유효성 검사와 타입 안정성을 동시에 확보 가능
- ```loginSchema```, ```signupSchema``` 등을 정의해 각 입력 폼의 유효성 조건을 선언
- react-hook-form의 ```zodResolver()```와 함께 사용해 입력값을 실시간으로 검증
- 비밀번호 확인(confirmPassword)처럼 입력값 간 관계 검증도 가능

## 이번 프로젝트에서 토큰 관리를 어떻게 할 예정인지, 그리고 왜 그런 방법을 선택했는지에 대해 설명해 주세요.
**클라이언트**
- 로그인 요청
- Access Token은 메모리나 쿠키에 저장=
- Refresh Token은 쿠키로만 보관 (httpOnly)
- 모든 API 요청 시 Authorization: Bearer ... 사용

**서버**
- Access Token 검증 → 통과 시 응답
- 만료 시 Refresh Token 검증 → 새 Access Token 발급
- 로그아웃 시 Refresh Token 무효화 (DB/Redis 등)

2025314242 committed:
Initial commit from Create Next App

2025314242 committed:
chore: initial setting

2025314242 committed:
chore: @svgr/webpack

2025314242 committed:
chore: initial setting

library, plugin ...

2025314242 committed:
design initial design

fonts, icons, favicon

2025314242 committed:
design: initial design

fonts, icons, favicon

2025314242 committed:
feat: initial project structure

2025314242 committed:
Merge branch 'main' into feat/initial-structure

2025314242 committed:
feat: initial project structure

app directory, components, libs, stores, styles

2025314242 committed:
docs: merge readme

only1Ksy committed:
chore: modify vote layout

only1Ksy committed:
design: vote button design

only1Ksy committed:
feat: vote store

only1Ksy committed:
feat: store vote info by zustand

only1Ksy committed:
feat: separate part / demo page UI

only1Ksy committed:
feat: change cast to slug folder

only1Ksy committed:
feat: modify header component

only1Ksy committed:
feat: cast page candidate card components

only1Ksy committed:
design: cast page components design detail

only1Ksy committed:
fix: modify demo to team

only1Ksy committed:
fix: modify vote page title

only1Ksy committed:
feat: vote cast page onSubmit Button function

only1Ksy committed:
chore: add icons for vote aggregate page

only1Ksy committed:
feat: vote aggregate page

only1Ksy committed:
fix: modify Select vote button logic

only1Ksy committed:
chore: pull modified code from feat/vote

only1Ksy committed:
fix: modify vote aggregate page title

only1Ksy committed:
feat: result page basic structure

only1Ksy committed:
design: modify pages design details

only1Ksy committed:
feat: add result show button in go to vote button component

only1Ksy committed:
feat: add crown to result page

only1Ksy committed:
design: apply hover color to buttons

only1Ksy committed:
design: modify design details of buttons

only1Ksy committed:
fix: modify header logic for result page

only1Ksy committed:
Merge branch 'feat/vote-result' into feat/vote-modal

only1Ksy committed:
feat: logout modal basic structure

only1Ksy committed:
design: change modal bg color

only1Ksy committed:
Merge pull request #4 from only1Ksy/feat/vote

feat: 투표 기능 구현, 기본 페이지 레이아웃 수정

only1Ksy committed:
Merge pull request #5 from only1Ksy/feat/vote-type-cast

feat: 타입캐스트 페이지 구조 및 투표 기능 구현

only1Ksy committed:
Merge pull request #6 from only1Ksy/feat/vote-aggregate

feat: aggregate 페이지

only1Ksy committed:
Merge pull request #7 from only1Ksy/feat/vote-result

feat: 결과 페이지 구현

only1Ksy committed:
Merge pull request #8 from only1Ksy/feat/vote-modal

feat: 로그아웃 모달 구현

2025314242 committed:
chore: zod, react-hook-form

2025314242 committed:
feature: login, sign-up page

2025314242 committed:
chore: api-setting

chore: zod, react-hook-form

2025314242 committed:
feature: login, sign-up page

feature: login, sign-up page

2025314242 committed:
Merge pull request #1 from Team-Promesa/main

fork

2025314242 committed:
feat: add href to back component

2025314242 committed:
feat: add href to back component

feat: add href to back component

2025314242 committed:
Merge pull request #2 from Team-Promesa/main

merge

chaeyoungwon commented on src/stores/vote-store.ts:
```diff
@@ -0,0 +1,43 @@
+      voteType: null,
+      selectedFrontend: null,
+      selectedBackend: null,
+      selectedTeam: null,
```
이 부분은 initialState로 분리해서 여기선 ...initialState 형태로 작성하고,
아래 reset에서는 set(initialState)로 초기화하는 방식으로 리팩토링해도 좋을 것 같습니다!

chaeyoungwon commented on src/libs/auth.ts:
stores 폴더가 아니라 libs 내에 저장하신 이유가 있나요??

2025314242 replied:
boilerplate 참고해서 폴더 구조 작성하다가 좀 꼬인 것 같네요 ㅠ-ㅠ 추후에는 lib/stores 이렇게 진행할 것 같습니다

chaeyoungwon commented on src/components/features/vote/result-grid.tsx:
```diff
@@ -0,0 +1,26 @@
+
+import ResultCard from './result-card';
+
+type ResultItem = {
+  name: string;
+  votes: number;
+};
```
ResultItem 타입은 따로 타입 파일로 분리해서 result-card에서도 import해서 사용하는 방식도 괜찮을 것 같아요!

chaeyoungwon commented on src/components/features/vote/header.tsx:
```diff
@@ -0,0 +1,48 @@
+
+  if (activeIndex === -1) return null;
+
+  const totalSteps = 3;
+  const sepLen = 5;
```
TOTAL_STEPS, SEPARATOR_COUNT는 바깥에 상수로 분리해두면 가독성이나 유지보수 측면에서 더 좋을 것 같아요!

chaeyoungwon commented on src/components/features/vote/header.tsx:
```diff
@@ -0,0 +1,48 @@
+
+export default function Header() {
+  const pathname = usePathname();
+  const segments = pathname.split('/');
+
+  const current = segments[3];
+  const sub = segments[4];
```
const [, , , current, sub] = pathname.split('/')처럼 바로 분리해주는 방법도 있더라고요! 코드가 조금 더 간결해질 수 있을 것 같아요 :)

chaeyoungwon commented on src/components/features/vote/header.tsx:
```diff
@@ -0,0 +1,48 @@
+  const current = segments[3];
+  const sub = segments[4];
+
+  const activeIndex = (() => {
+    if (current === 'list') return 0;
+    if (['front', 'back', 'team', 'aggregate', 'loading'].includes(current) && sub !== 'result') return 1;
+    if (sub === 'result') return 2;
+    return -1;
+  })();
```
getActiveIndex라는 함수를 따로 만들어서 받아와도 좋을 거 같네용

chaeyoungwon commented on src/components/features/vote/header.tsx:
```diff
@@ -0,0 +1,48 @@
+  return (
+    <div className="mt-6 flex items-center justify-between">
+      {Array.from({ length: totalSteps }).map((_, stepIdx) => (
+        <React.Fragment key={`step-${stepIdx}`}>
```
저는 <></>를 못 쓰는 상황에서는 습관적으로 의미 없는 `<div></div>` 로 감싸는 편인데,, React.Fragment를 활용해주신 게 좋은 거 같네용

seoyeon5117 replied:
헉 저도 배워갑니다,,,

chaeyoungwon commented on src/components/features/vote/candidate-grid.tsx:
```diff
@@ -0,0 +1,14 @@
+  const columns = list.length <= 5 ? 1 : 2;
+
+  return (
+    <div className={`grid grid-rows-5 gap-x-4 gap-y-3 ${columns === 1 ? 'grid-cols-1' : 'grid-cols-2'}`}>
```
코드 전반적으로 조건부 스타일링이 꽤 있는데, clsx 같은 라이브러리를 활용해보셔도 좋을 것 같아요!

chaeyoungwon commented on src/components/features/vote/candidate-card.tsx:
```diff
@@ -0,0 +1,45 @@
+  };
+
+  return (
+    <button
+      onClick={handleClick}
+      className={
+        'bg-grayscale-04 text-body-01 hover:bg-neutral-01-tp relative flex h-13.5 cursor-pointer items-center justify-center border-2 border-black transition-colors'
+      }
+    >
+      {name}
+      {isSelected && (
+        <>
+          {/* 반투명 레이어 */}
+          <div className="bg-neutral-01-tp absolute inset-0 z-10 rounded-sm" />
+
+          {/* 체크 아이콘 */}
+          <button className={`bg-neutral-01 top-50% right-50% absolute z-20 h-8.5 w-8.5 rounded-full border-2`}>
+            <span className="text-xl font-bold text-gray-700">✓</span>
+          </button>
+        </>
+      )}
+    </button>
```
선택됐을 때 나오는 체크 버튼이 `<button> `안에 또 `<button>`으로 들어가 있는데 HTML 구조상 권장되지 않는 방식이라고 하네요!
체크 아이콘을 버튼으로 쓰고 싶을 때는, 외부를 `<div role="button">`으로 바꾸는 방식도 고려해보면 좋을 것 같아요!

chaeyoungwon commented on src/components/features/auth/input-form.tsx:
오 react-hook-form 쓰셨군요 저도 다음에 리팩토링할 때 써봐야겠어요 ..

2025314242 replied:
사용하실 때 @hookform/resolvers 이 라이브러리도 설치하시면 도움되실거에요!

chaeyoungwon commented on src/components/features/auth/button.tsx:
```diff
@@ -0,0 +1,29 @@
+import Link from 'next/link';
+
+type ButtonProps =
+  | {
+      text: string;
+      href: string;
+      onClick?: never;
+    }
+  | {
+      text: string;
+      onClick: () => void;
+      href?: never;
+    };
```
union으로 타입을 분리하신 점이 인상깊네용

chaeyoungwon commented on src/components/features/auth/button.tsx:
```diff
@@ -0,0 +1,29 @@
+
+export default function Button({ text, href, onClick }: ButtonProps) {
+  const content = (
+    <div className="bg-grayscale-00-black hover:bg-neutral-02 group flex h-12 cursor-pointer items-center justify-center rounded-[50] drop-shadow-[4px_6px_8px_rgba(0,0,0,0.25)]">
```
이 shadow 스타일이 여러 곳에서 쓰이고 있는 것 같아서, css에 변수로 따로 정의해두면 관리하기 더 편할 것 같아요!

chaeyoungwon commented on src/components/common/indicator.tsx:
```diff
@@ -0,0 +1,23 @@
+        onClick={() => setShowLogout(true)}
+        className="absolute bottom-5 z-999 flex w-full cursor-pointer justify-center gap-2"
+      >
+        <div className="bg-grayscale-00-black h-3 w-3 rounded-full"></div>
+        <div className="bg-grayscale-00-black h-3 w-3 rounded-full"></div>
+        <div className="bg-grayscale-00-black h-3 w-3 rounded-full"></div>
```
이 부분은 자식 요소가 없는 구조라 self-closing 형태로 작성해도 될 것 같아요!

2025314242 replied:
음 그러네요 감사합니다 !

chaeyoungwon commented on src/app/(auth)/login/page.tsx:
```diff
@@ -0,0 +1,47 @@
+  return (
+    <div className="flex h-full flex-col px-10">
+      <div className="mt-8 mb-15 flex items-center">
+        <Back href="/" />
```
백 버튼을 따로 빼서 각 페이지 주소를 직접 넣어주신 걸로 보이는데,
혹시 router.back() 대신 명시적인 경로를 사용하신 이유가 있을까요?
새로고침이나 특정 상황에서도 안정적으로 이동시키기 위한 의도일지 궁금합니다!

2025314242 replied:
넵 브라우저에 있는 뒤로가기 버튼이랑 좀 차별점 두기 위해서 이렇게 설정했습니다 !

chaeyoungwon reviewed (commented):
과제하느라 수고 많으셨습니다 ~~ 피드백 남겨봅니다 🙇‍♀️

xseojungx commented on src/app/(auth)/login/page.tsx:
지금 따로 회원 가입을 안 해도 로그인이 되는데 api 연결 때문에 아직 user auth 부분을 구현 못 하신건가요?

xseojungx commented on src/app/vote/[vote-type]/list/page.tsx:
vote 아래 list 폴더가 두 번 있는데 설계상 일부러 다르게 둔 건가요? 
/vote/list 전체 투표
/vote/[vote-type]/list/page.tsx  특정 타입 투표라면 라이팅 주소를 조금 다르게 하는 등의 차이를 주면 좋을 것 같아요!

xseojungx commented on src/app/vote/[vote-type]/[cast]/page.tsx:
```diff
@@ -0,0 +1,71 @@
+    return TEAMS;
+  };
+
+  const getTitle = () => {
```
여러 페이지에서 비슷하게 반복되는 것 같은데 유틸 함수에 따로 빼놔도 좋을 것 같아요!

xseojungx commented on src/app/vote/[vote-type]/[cast]/page.tsx:
```diff
@@ -0,0 +1,71 @@
+  const onSubmit = () => {
+    let selected: string | null = null;
+
+    if (type === 'front') selected = selectedFrontend;
```
여기서 store을 사용해서 해당 정보를 저장하는 이유는 어떻게되나요?

only1Ksy replied:
처음에 컴포넌트를 해당 값으로 제어하다가 수정하는 과정에서 빼 버렸는데 삭제를 안 했습니당...

xseojungx commented on src/stores/vote-store.ts:
```diff
@@ -0,0 +1,43 @@
+  reset: () => void;
+}
+
+export const useVoteStore = create<VoteStore>()(
```
개인적인 생각이지만 프로젝트 규모에 비해서 store에 저장하는 요소들이 많은 것 같아요! store 사용하면 좋지만 이게 오히려 메모리 사용으로 인한 성능 저하를 일으킬 수 있을 것 같아서 props로 전달해도 depth가 너무 깊지 않거나 문제가 되지 않는다면 그냥 props로 해도 좋을 것 같습니다!

seoyeon5117 commented on src/app/(auth)/page.tsx:
```diff
@@ -0,0 +1,56 @@
+              </p>
+            </div>
+            <div className="relative flex-[5]">
+              <VoteIcon className="text-grayscale-00-black-tp absolute top-2/5 drop-shadow-[4px_6px_8px_rgba(0,0,0,0.25)]" />
```
그림자가 반복해서 쓰이니 global로 따로 빼줘도 좋을 것 같아요

seoyeon5117 commented on src/app/(auth)/page.tsx:
```diff
@@ -0,0 +1,56 @@
+          <>
+            <Button href="/login" text="Join us" />
+            <Link href="/sign-up">
+              <div className="mt-[14px] flex justify-center">
```
px-10은 rem인데 mt-[14px]은 px이라 rem으로 바꾸는게 좋아보입니다!

xseojungx commented on src/components/common/modal.tsx:
import { createPortal } from 'react-dom';
이거 사용해서 모달 처리하면 좋을 것 같아요!

xseojungx reviewed (commented):
수고하셨습니다! 이번주 구현 요구사항보다 많이 개발하시느라 수고하셨어요

seoyeon5117 commented on src/components/features/auth/input-form.tsx:
```diff
@@ -0,0 +1,28 @@
+        <label className="text-accent-brown">{label}</label>{' '}
+        {error && <p className="text-[10px] text-red-500">{error.message}</p>}
+      </div>
+      <input {...register(name)} type={type} className="text-body-02 outline-none" />
```
label이랑 input 같이 사용하실 때는 id랑 htmlFor 함께 사용하시면 좋을 것 같습니당

seoyeon5117 commented on src/components/features/vote/candidate-card.tsx:
```diff
@@ -0,0 +1,45 @@
+          <div className="bg-neutral-01-tp absolute inset-0 z-10 rounded-sm" />
+
+          {/* 체크 아이콘 */}
+          <button className={`bg-neutral-01 top-50% right-50% absolute z-20 h-8.5 w-8.5 rounded-full border-2`}>
```
top-50%는 top-1/2로도 쓸 수 있습니당

seoyeon5117 commented on src/models/auth.ts:
저도 로직 분리해봐야겠네요

seoyeon5117 commented on src/stores/vote-store.ts:
```diff
@@ -0,0 +1,43 @@
+        }),
+    }),
+    {
+      name: 'vote-storage', // localStorage에 저장될 key 이름
```
만약 localStorage를 사용하신다면 key 값을 상수로 따로 빼서 저장해도 좋을 것 같네요!

seoyeon5117 reviewed (commented):
과제하시느라 수고하셨습니다!!

2025314242 committed:
fix: modify path alias

fix: modify path aliasing

feat: add href to back component

fix: modify path alias - remaining

2025314242 committed:
chore: install axios

2025314242 committed:
Merge remote-tracking branch 'upstream/master' into main

2025314242 committed:
fix: merge pull request #3 from Team-Promesa/main

Forward to sub repository

2025314242 committed:
fix: merge pull request #13 from 2025314242/main

Forward to upstream repository

2025314242 committed:
feat: auth store zustand

2025314242 committed:
Merge branch 'main' of graduate:2025314242/ceos.week7to8.next-vote-21th into main

2025314242 committed:
fix: remaining path aliasing error

2025314242 committed:
fix: deploy error

1. zustand useAuthStore
2. remaining path aliasing error

2025314242 committed:
feat: auth api and user hook

- auth api (login, signup, logout)
- user hook

2025314242 committed:
feat: adapt auth api to auth pages

2025314242 committed:
fix: indicator tag typo

2025314242 committed:
fix: indicator tag typo

2025314242 committed:
feat: auth api and user hook

2025314242 committed:
Merge pull request #17 from 2025314242/feat/auth-page-with-api

feat: adapt auth api to auth pages

2025314242 committed:
refactor: member data

2025314242 committed:
Merge pull request #18 from 2025314242/refactor/member-data

refactor: member data

2025314242 committed:
feat: add part/team property selection

2025314242 committed:
Merge pull request #19 from 2025314242/feat/auth-part-team

feat: add part/team property selection

2025314242 committed:
feat: middleware

2025314242 committed:
Merge pull request #20 from 2025314242/feat/middleware

로그인 여부에 따라서 리다이렉션

2025314242 committed:
feat: logout modal

2025314242 committed:
Merge pull request #21 from 2025314242/feat/logout-modal

로그아웃 모달에 실제로 동작하도록 추가함

2025314242 committed:
fix: useUser hook

2025314242 committed:
Merge pull request #22 from 2025314242/fix/use-user-hook

useUser hook에서 getUser(SSR 전용) 사용해서 에러뜸
-> api/auth/me/route.ts 통해서 호출하도록 수정

2025314242 committed:
docs: notion url

2025314242 committed:
Merge branch 'main' of graduate:2025314242/ceos.week7to8.next-vote-21th into main

2025314242 committed:
Merge pull request #23 from 2025314242/main

docs

2025314242 committed:
feat: backend API

2025314242 committed:
Merge pull request #24 from 2025314242/feat/backend-api

backend API랑 연결
연결하면서 코드 리팩토링 진행 (최대한 원본 코드 유지하려고 함)
문제점:
backend API가 dev 환경에서 적용이 안됨 (아마 next.config.ts rewrites not working error인듯. next JS 최신 버전 오류)
deployment 시 적용될지 테스트

2025314242 committed:
fix: remove NEXT_PUBLIC_BASE_URL

2025314242 committed:
Merge pull request #25 from 2025314242/fix/backend-url

NEXT_PUBLIC_BASE_URL 제거
BACKEND_URL 추가

2025314242 committed:
fix: logged in

2025314242 committed:
Merge pull request #26 from 2025314242/fix/logged-in

fix: logged in

2025314242 committed:
fix: refresh token

2025314242 committed:
Merge pull request #27 from 2025314242/fix/refresh-token

fix: refresh token

2025314242 committed:
test: remove console log

2025314242 committed:
fix: axios error

2025314242 committed:
feat: api-auth-me

2025314242 committed:
design: join us -> login

2025314242 committed:
feat: backend requirements

2025314242 committed:
Merge pull request #28 from 2025314242/feat/requirements

Feat/backend requirements

2025314242 committed:
feat: backend requirements v2

2025314242 committed:
Merge pull request #29 from 2025314242/feat/requirements-v2

Feat/requirements-v2

2025314242 committed:
fix: crown only if voteCount > 0

2025314242 committed:
Merge pull request #30 from 2025314242/fix/crown

fix: crown only if voteCount > 0

2025314242 committed:
fix: z-999 to logout modal
